### DISCUSSION TITLE
delete from table with foreign keys

### DISCUSSION THREAD

**Discussion by EdgarsLv** [supabase/supabase#6524] [Questions] [answered]
I want to delete user from app and users table.
If user doesn't have foreign key constraints i can delete user like this(while not sure if it's the correct way to do it):
![image](https://user-images.githubusercontent.com/56837461/164647017-0d197a8f-c2fb-436e-b224-cc24cf1a8534.png)

but if he has foreign keys in other tables i can't delete anymore and getting foreign key constraint error.

1. How to delete user from auth and user table with foreign keys if i don't want to delete all related data?

2. And how to delete if i want to delete user completely with all hes related(example, all messages from messages table and  all conversations from conversations table) data?

**Comment by GaryAustin1** [comment, marked as answer]
1)  You can alter or setup your foreign key constraint and not do the default or "on delete cascade" and instead do "on delete set null".  You have to allow nulls in your foreign key column.  Then when you do auth.api.deleteUser() the data will be left around with null as user id.  If you have a public user/profile table you could then change the user's name to Guest or Former User or something if desired.
See 5.4.5. Foreign Keys in https://www.postgresql.org/docs/current/ddl-constraints.html

2) If you have add "on delete cascade" as part of the foreign key constraint then when the user is deleted the row with the key pointed to is deleted.  Also any foreign keys point to that row will have their rows deleted if they have "on delete cascade" set.
